python/cudf/cudf/core/groupby/groupby.py

```diff
@@ -590,6 +571,47 @@ def rolling(self, *args, **kwargs):
         return cudf.core.window.rolling.RollingGroupby(self, *args, **kwargs)
 
 
+# Set of valid groupby aggregations that are monkey-patched into the GroupBy
+# namespace.
+_VALID_GROUPBY_AGGS = {
+    "count",
+    "sum",
+    "idxmin",
+    "idxmax",
+    "min",
+    "max",
+    "mean",
+    "var",
+    "std",
+    "quantile",
+    "median",
+    "nunique",
+    "collect",
+}
+
+
+# Dynamically bind the different aggregation methods.
+for key in _VALID_GROUPBY_AGGS:
+
+    def _agg_func_name_with_args(self, func_name, *args, **kwargs):
```

Comment:
Can we not move this to outside the loop?

Reply:
Yes, we definitely could. The reason I hadn't done that is because I was instead trying to modify this function to not require the function name as input, but to instead just use the `key` directly in the `getattr` call inside. For some reason that didn't work, though: some of the `test_groupby.py` tests failed and the error wasn't something obvious, so I didn't take the time to dig further. Any ideas why that might be? My guess is something due to when the local variable is bound makes `getattr` pull the wrong attribute from `x` at runtime, but I'm not really sure.

Reply:
In any case, perhaps moving it outside the loop and calling it is the better option anyway.

Reply:
> Any ideas why that might be? 

Looks like a case of [late binding](https://docs.python-guide.org/writing/gotchas/#late-binding-closures)?

Reply:
Agree, I like what we have now

Reply:
Ah yes, thanks for finding a good link. I dealt with this once [a long time ago](https://bitbucket.org/glotzer/hoomd-blue/pull-requests/222) (can't believe it's been five years...) and forgot the exact rules. But good, the new version is a definite improvement. I'll work on tests next.